EC-CUBE 4.0ブランチに移植

nanasess opened the pull request:
## #76からの変更点
- PurchaseFlow を使用するよう修正
- Order::setOrderCompleteMailMessage() を使用してメールの差し込みをするよう修正
- クーポン値引明細の集計方法を変更

## 値引の計算方法について

支払い時、クーポンを商品券のように扱うか、売上代金の値引として扱うかによって消費税の計算方法が変わってくる
- **商品券方式:** 支払い金額を代替する金券なので税込みに対して値引きする
- **商品値引方式:** 商品の値段を値引くものなので税抜に対して値引きする

EC-CUBE4のポイント機能は、商品券方式を採用しており、ポイント機能との併用も考慮して、クーポンプラグインも商品券方式とする

ポイント機能と併用した場合は、ポイント値引後の価格に対して適用する

### 値引率の場合の計算方法

- 明細ごとに税込の値引額を集計し、クーポン値引の明細を生成する(税込価格、税率0%、課税)
- 軽減税率適用により税率が混在する場合もあるため、税込価格、税率0%で明細を生成する
- ポイント値引後の合計金額に値引を適用する

### 値引額の場合の計算方法

- クーポンで設定した価格で、クーポン値引の明細を生成する(税込価格、税率0%、課税)
- ポイント値引後の合計金額に値引を適用する

## ポイント機能との併用について

- ポイント利用時の加算ポイントについては、クーポンは考慮しない(クーポンを利用しても加算ポイントは変わらない)
  - *ポイント利用時、ポイント利用分の加算ポイントが控除される仕様は1系からの歴史的な理由*
- ポイント利用後の合計金額に対してクーポン値引を適用する
- 合計金額よりクーポン値引額が上回っても利用可能(おつりは出ない)

## 現状の課題

商品券方式を採用することにより、以下の課題は解決

- ~ポイントとの併用ができない。 Processor の優先順位がポイントの方が高く、値引率で集計できなくなるため。 purchaseflow.yaml を手動で設定することで回避可能。~
- ~軽減税率に対応していない。値引率で集計する際、明細ごとに割引価格を集計するが、クーポン値引明細は1行なので、複数の税率に対応できない。クーポン明細を複数行で出力にすることで回避可能~

## TODO

- [x] クーポンの妥当性検証を CouponProcessor に移植する
- [x] Event::onOrderEditComplete() を CouponStateProcessor に移植する
- [x] ユニットテスト

## 参考

[マイレージサービスに代表されるポイント制に係る税務上の取扱い －法人税・消費税の取扱いを中心に－](https://www.nta.go.jp/about/organization/ntc/kenkyu/ronsou/58/01/hajimeni.htm)

lqdung-lockon committed:
First commit

lqdung-lockon committed:
Change text

lqdung-lockon committed:
Fix text

lqdung-lockon committed:
Fix the admin new coupon function

lqdung-lockon committed:
Add coupon new/edit layout

lqdung-lockon committed:
Fix shopping input coupon

lqdung-lockon committed:
Add mail template + order edit/complete

lqdung-lockon committed:
Add message translate (50%)

lqdung-lockon committed:
change message (done)

lqdung-lockon committed:
Add order edit event

lqdung-lockon committed:
Add unit test for coupon

lqdung-lockon committed:
Remove comment

lqdung-lockon committed:
Fix comment

lqdung-lockon committed:
Add delete modal

lqdung-lockon committed:
Remove comment

lqdung-lockon committed:
Fix message

lqdung-lockon committed:
Add copy block function

lqdung-lockon committed:
Fix message translate

lqdung-lockon committed:
Fix phpCS

lqdung-lockon committed:
Fix php unit

lqdung-lockon committed:
Remove deprecated feature

lqdung-lockon committed:
Mark skip some test case

lqdung-lockon committed:
Merge from upstream

lqdung-lockon committed:
Fix new style

lqdung-lockon committed:
Move to @template

lqdung-lockon committed:
Change visible to bool

lqdung-lockon committed:
Build travis scenario

lqdung-lockon committed:
Fix name upper

lqdung-lockon committed:
Fix visible to bool

lqdung-lockon committed:
Fix phpcs

lqdung-lockon committed:
Change format message

lqdung-lockon committed:
Change magic number to constant

nanasess committed:
Change branch to 4.0

nanasess committed:
メールに追加するメッセージの実装を4.0に合わせる

nanasess committed:
[WIP]PurchaseFlow に移植

nanasess committed:
[WIP]管理画面側の Processor を追加

nanasess committed:
値引明細の追加処理を CouponProcessor に実装

nanasess committed:
不要な処理を削除

nanasess committed:
Apply php-cs-fixer

nanasess committed:
Merge remote-tracking branch 'upstream/experimental/sf' into 4-support

nanasess committed:
Remove ShoppingService

nanasess committed:
Fix parameters

nanasess committed:
config.yml から composer.json に変更

nanasess committed:
Fix namespace to Coupon4

nanasess committed:
Fix Page

nanasess committed:
Fix directories

nanasess committed:
Fix namespace

nanasess committed:
Fix nav

nanasess committed:
Fix phpDoc comment

nanasess committed:
Fix DI

nanasess committed:
クーポンの妥当性検証を CouponProcessor に移植

nanasess committed:
Event::onOrderEditComplete() を CouponStateProcessor に移植する

nanasess committed:
Fix travis.yml

nanasess committed:
Fix travis.yml

nanasess committed:
Fix travis.yml

nanasess committed:
Fix travis.yml

nanasess committed:
Fix branch

nanasess committed:
Fix PLUGIN_CODE

nanasess committed:
Fix testcase

nanasess committed:
Fix Testcase

nanasess committed:
Fix Testcase for Frontend

nanasess committed:
Fix enable/disable and delete

nanasess committed:
Refactoring

nanasess committed:
初回に送料計算ができなくなってしまう不具合修正

nanasess committed:
商品購入と Myページのテストを動くよう修正

nanasess committed:
Fix jobs to coverage

nanasess committed:
Add OrderControllerTest

nanasess committed:
Fix jobs to coverage

nanasess committed:
Add eccube_setup

nanasess committed:
Fix coveralls commands

nanasess committed:
受注編集時のテスト

nanasess committed:
use phpdbg

nanasess committed:
Fix phpunit.xml.dist

nanasess committed:
クーポン値引明細の属性を変更

nanasess committed:
incomplete のマークを削除

nanasess committed:
Fix appveyor

kiy0taka commented on Service/PurchaseFlow/Processor/CouponProcessor.php:
```diff
@@ -0,0 +1,316 @@
+        $Customer = $itemHolder->getCustomer();
+        if (!$Customer->getId() && $Coupon->getCouponMember()) {
+            $this->throwInvalidItemException(trans('plugin_coupon.front.shopping.member'));
+            $this->clearCoupon($itemHolder);
```
145行目で例外がスローされるのでこの行は実行されません。

nanasess replied:
修正しました
https://github.com/EC-CUBE/coupon-plugin/commit/c03f90a4077f90259d8b40dde3bac06747c3b0af

kiy0taka commented on Service/PurchaseFlow/Processor/CouponProcessor.php:
```diff
@@ -0,0 +1,316 @@
+        $Coupon = $this->couponRepository->findActiveCoupon($CouponOrder->getCouponCd());
+        if (!$Coupon) {
+            $this->throwInvalidItemException('クーポンが取得できませんでした', null, true);
+            $this->clearCoupon($itemHolder);
```
138行目で例外がスローされるのでこの行は実行されません。

nanasess replied:
修正しました
https://github.com/EC-CUBE/coupon-plugin/pull/77/commits/13d1b2a1b4479f899cfd417c2adfc5acbbb9faa3

kiy0taka commented on Service/PurchaseFlow/Processor/CouponProcessor.php:
```diff
@@ -0,0 +1,316 @@
+        $couponProducts = $this->couponService->existsCouponProduct($Coupon, $itemHolder);
+        if (count($couponProducts) == 0) {
+            $this->throwInvalidItemException(trans('plugin_coupon.front.shopping.couponusetime'), null, true);
+            $this->clearCoupon($itemHolder);
```
152行目で例外がスローされるのでこの行は実行されません。

nanasess replied:
修正しました
https://github.com/EC-CUBE/coupon-plugin/commit/c03f90a4077f90259d8b40dde3bac06747c3b0af

kiy0taka commented on Service/PurchaseFlow/Processor/CouponProcessor.php:
```diff
@@ -0,0 +1,316 @@
+        if (!$checkLowerLimit) {
+            $message = trans('plugin_coupon.front.shopping.lowerlimit', ['lowerLimit' => number_format($lowerLimit)]);
+            $this->throwInvalidItemException($message, null, true);
+            $this->clearCoupon($itemHolder);
```
159行目で例外がスローされるのでこの行は実行されません。

nanasess replied:
修正しました
https://github.com/EC-CUBE/coupon-plugin/pull/77/commits/c03f90a4077f90259d8b40dde3bac06747c3b0af

kiy0taka commented on Service/PurchaseFlow/Processor/CouponProcessor.php:
```diff
@@ -0,0 +1,316 @@
+        $checkCouponUseTime = $this->couponRepository->checkCouponUseTime($CouponOrder->getCouponCd());
+        if (!$checkCouponUseTime) {
+            $this->throwInvalidItemException(trans('plugin_coupon.front.shopping.couponusetime'), null, true);
+            $this->clearCoupon($itemHolder);
```
167行目で例外がスローされるのでこの行は実行されません。

nanasess replied:
修正しました
https://github.com/EC-CUBE/coupon-plugin/pull/77/commits/c03f90a4077f90259d8b40dde3bac06747c3b0af

kiy0taka commented on Controller/CouponShoppingController.php:
```diff
@@ -0,0 +1,259 @@
+    /**
+     * @var ShoppingService
+     */
+    private $shoppingService;
```
`ShoppingService` は存在しません。

kiy0taka commented on Controller/CouponShoppingController.php:
```diff
@@ -0,0 +1,259 @@
+                if ($this->isGranted('ROLE_USER')) {
+                    $Customer = $this->getUser();
+                } else {
+                    $Customer = $this->shoppingService->getNonMember($this->sessionKey);
```
`ShoppingService` は存在しません。

kiy0taka commented on Controller/CouponShoppingController.php:
```diff
@@ -0,0 +1,259 @@
+                $error = false;
+                // クーポン情報を取得
+                /* @var $Coupon Coupon */
+                $Coupon = $this->couponRepository->findActiveCoupon($formCouponCd);
```
この辺の妥当性検証は`CouponProcessor`で実装されているので`PurchaseFlow`を利用してください。

nanasess replied:
ここのエラーハンドリングは、クーポン入力フォームにエラーメッセージを表示したいので、ここで実装しています。
(`CouponProcessor` で実装すると、注文確認画面にエラーメッセージを表示してしまうため)

kiy0taka replied:
了解しました。ここはこれで行きましょう。

kiy0taka commented on Service/PurchaseFlow/Processor/CouponProcessor.php:
```diff
@@ -0,0 +1,316 @@
+            $this->clearCoupon($itemHolder);
+        }
+
+        $lowerLimit = $Coupon->getCouponLowerLimit();
```
このコードは156行目の上にある方が可読性が高くなります。

nanasess replied:
修正しました
https://github.com/EC-CUBE/coupon-plugin/pull/77/commits/93d781da7ccd45fc1a55b3051a6963e17621d98a

kiy0taka reviewed (changes requested):
ライセンス表記の修正もお願いします。

nanasess committed:
ShoppingService を OrderHelper に変更

nanasess committed:
エラーハンドリングの修正

nanasess committed:
可読性の向上

nanasess committed:
不要なコメントの削除

nanasess committed:
エラーハンドリングの修正

nanasess committed:
Apply php-cs-fixer

nanasess commented:
ライセンス表記修正しました

nanasess committed:
form_label にもエラーメッセージが表示されてしまうのを修正

nanasess committed:
CouponProcessor を検索条件に削除するよう修正

nanasess committed:
非会員購入のテストケース追加

nanasess committed:
Fix message codes

nanasess committed:
Delete confirmation message

nanasess committed:
Use PriceType

nanasess committed:
Fix complete mail message

nanasess committed:
値引き対象の商品全てに適用されないのを修正

nanasess committed:
Delete confirmation message

nanasess committed:
MySQL で 非表示の注文明細ができてしまうのを修正

createOrder で生成した受注に新たに明細を追加するのではなく、
CouponDetail から取得した ProductClass を createOrder に渡すよう修正

nanasess committed:
Use xenial build environment
